### PR TITLE
server: always create a liveness record before starting up

### PR DESCRIPTION
**Pull request by irfansharif**
Previously it used to be the case that it was possible for a node to be
up and running, and for there to be no corresponding liveness record for
it. This was a very transient situation as liveness records are created
for a given node as soon as it out its first heartbeat. Still, given
that this could take a few seconds, it lent to a lot of complexity in
our handling of node liveness where we had to always anticipate the
possibility of there being no corresponding liveness record for a given
node (and thus creating it if necessary).

Having a liveness record for each node always present is a crucial
building block for long running migrations (#48843). There the intention
is to have the orchestrator process look towards the list of liveness
records for an authoritative view of cluster membership. Previously when
it was possible for an active member of the cluster to not have a
corresponding liveness record (no matter how unlikely or short-lived in
practice), we could not generate such a view.

---

This is an alternative implementation for #53805. Here we choose to
manually write the liveness record for the bootstrapping node when
writing initial cluster data. For all other nodes, we do it on the
server-side of the join RPC. We're also careful to do it in the legacy
codepath when joining a cluster through gossip.

Release note: None

### CODE REVIEW

**Comment by irfansharif** [comment]
+cc @tbg, mind taking a look here as well? I'm still not sure if there's a simpler workaround for the --locality-advertise-addr issue, I feel like I must be missing something obvious. Though that said, writing to the store directly during bootstrap would work around that problem if nothing else is possible. See comments in the branch around epoch={0,2}, wanted your thoughts there.

**Comment by irfansharif** [comment]
We probably don't care about having the liveness `epoch=2` for new nodes, so I'll remove the "initialize at epoch=0" change here.

**Review by tbg — APPROVED** [review]
This came out well!

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-:-MGn6bV_1g5mz8OysMXu:b2kgb9w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[pkg/kv/kvserver/node_liveness.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn4sMJBHoiRe2mcF1q:-MGn4sMK7IBlrLHXpKl_:b-27csg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/kv/kvserver/node_liveness.go#L461)):*
> ```Go
> // records after starting up, and incrementing to epoch=2 when doing so, at
> // which point we'll set an appropriate expiration timestamp, gossip the
> // liveness record, and update our in-memory representation of it.
> ```

Add/reword that an existing liveness record would not be overwritten but instead an error returned from this method.
___
*[pkg/kv/kvserver/node_liveness_test.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn5hhq14BRHIMeNUdf:-MGn5hhr2attO9ivhIR6:b-j61z0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/kv/kvserver/node_liveness_test.go#L174)):*
> ```Go
> 	defer tc.Stopper().Stop(ctx)
> 
> 	// Verify liveness records exist for all nodes.
> ```

So at this point, something in `StartTestCluster` has waited for all nodes to become live (otherwise this test is just flaky)? Add a comment.
___
*[pkg/server/node.go, line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn6FDR8-PYj5njWVzC:-MGn65yvDljqJtcxgYqa:b-62ts7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/server/node.go#L383)):*
> ```Go
> 		// We're joining via gossip, so we don't have a liveness record for
> 		// ourselves yet. Let's create one while here.
> 		if err := n.storeCfg.NodeLiveness.CreateLivenessRecord(ctx, nodeID); err != nil {
> ```

Just reasoning this out to myself: we're not in the case in which "we are the cluster" because that only happens when this is the node that just got bootstrapped, but in that case it has a nodeID.
___
*[pkg/server/node.go, line 1160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn6NOCBrBA03mHnUJK:-MGn6NOCBrBA03mHnUJL:bfauo2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/server/node.go#L1160)):*
> ```Go
> 	// We create a liveness record here for the joining node while here. This
> 	// way nodes are always guaranteed to have a liveness record present before
> 	// fully starting up.
> ```

Add that this is required for long-running migrations correctness. (This might be a place where folks in the future might muck with this stuff and not know why it was the way it is now in the first place).

**Review by irfansharif — COMMENTED** [review]
TFTR! CI had failed on https://github.com/cockroachdb/cockroach/issues/54079, which is a bit strange (+ funny). I think something else is going on there, and I'm curious to know what. I was also missing an update to some logic test.

I'm a bit sad about not being able to use epoch=0, for no good reason. I think it's still possible to do so safely. I'll merge this when green, and might try lowering it again on `master` alone just to get back to the previous behavior.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-:-MGp7kNC-1xY5bghD5Ja:b-5ytyku)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/node_liveness.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn4sMJBHoiRe2mcF1q:-MGp7cqEBNp2eRH7zj_b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/kv/kvserver/node_liveness.go#L461)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Add/reword that an existing liveness record would not be overwritten but instead an error returned from this method.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/node_liveness_test.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn5hhq14BRHIMeNUdf:-MGp7fL95DdOuFhOTlWF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/kv/kvserver/node_liveness_test.go#L174)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

So at this point, something in `StartTestCluster` has waited for all nodes to become live (otherwise this test is just flaky)? Add a comment.
</blockquote></details>

Done.
___
*[pkg/server/node.go, line 1160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGn6NOCBrBA03mHnUJK:-MGp7iEFAHDqbsECaugQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b976f92f9a32989ef0b0f305ef29e1a0ee4d27c5/pkg/server/node.go#L1160)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Add that this is required for long-running migrations correctness. (This might be a place where folks in the future might muck with this stuff and not know why it was the way it is now in the first place).
</blockquote></details>

Done.

**Comment by irfansharif** [comment]
> I'm a bit sad about not being able to use epoch=0, for no good reason. I think it's still possible to do so safely. I'll merge this when green, and might try lowering it again on master alone just to get back to the previous behavior.

Hm, looks like I'll actually need to go with the epoch=0 initial write. Here's why: the existing node liveness tests use multiTestContext and all its quirks (I made use of TestCluster in the test I just added). One of the quirks is that it makes do without using Servers. Which also means it makes do without using the join RPC added in #52526. As a result, all nodes except for the bootstrap node will not actually have a liveness record persisted in KV. 

Let's say we write our initial liveness record at epoch=1. For the bootstrap node, given the NodeLiveness component starts off knowing about this liveness record, we'll increment it to epoch=2 during our first heartbeat.

https://github.com/cockroachdb/cockroach/blob/d0ead9a5814b5ac329bbf3a5738b4331c1ebf90f/pkg/kv/kvserver/node_liveness.go#L761-L763

For all other nodes, given we don't already have a liveness record persisted, after our first heartbeat we'll find ourselves at epoch=1.

https://github.com/cockroachdb/cockroach/blob/d0ead9a5814b5ac329bbf3a5738b4331c1ebf90f/pkg/kv/kvserver/node_liveness.go#L756-L758

The tests that want to assert that epoch=1 for all liveness records, without having seen restarts, will not be too happy about this discrepancy.

https://github.com/cockroachdb/cockroach/blob/d0ead9a5814b5ac329bbf3a5738b4331c1ebf90f/pkg/kv/kvserver/node_liveness_test.go#L152-L154

https://github.com/cockroachdb/cockroach/blob/d0ead9a5814b5ac329bbf3a5738b4331c1ebf90f/pkg/kv/kvserver/node_liveness_test.go#L613-L615


This is another instance where having #8299 would have come in handy. I went ahead and kept the initial write to be at epoch=0 so we retain the previous epoch=1 pattern for very first start. That way at least the node liveness tests that make use of multiTestContext pass. As for the changes introduced in this PR, TestNodeLivenessAppearsAtStart sufficiently exercises those code paths.

---

I can try and follow up this month/next with migrating some of these tests over to use TestCluster (partially addressing #8299), but I'm inclined to not hold this PR up for it. I'll bors it once you take another look, @tbg.

**Comment by irfansharif** [comment]
> One of the quirks is that it makes do without using Servers. Which also means it makes do without using the join RPC added in #52526. As a result, all nodes except for the bootstrap node will not actually have a liveness record persisted in KV.

I suppose it's fine for me to manually write liveness records as part of multiTestContext.Start, which I just did. Doh. I still prefer starting off at epoch=0. I'm not too worried about the zero value of the liveness record because it isn't the zero value, the node ID is set. My motivation is primarily to retain epoch=1 semantics that we have today. But I can change it out again if asked.

**Comment by tbg** [comment]
I'm taking a look now, but already wanted to let you know that I'm fine writing at epoch zero. My initial inclination to avoid that was because you mentioned something tricky resulted from it (or so I remember). If that is not the case - perfect, epoch zero it is. Looking now through the code though.

**Review by tbg — APPROVED** [review]
> Reviewed 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[pkg/kv/kvserver/node_liveness.go, line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGrE42x0TpH7S8PtOoG:-MGrE42x0TpH7S8PtOoH:bbsd749) ([raw file](https://github.com/cockroachdb/cockroach/blob/64c3b57ebab29e59c19bf16f1c40bffc8591f8a4/pkg/kv/kvserver/node_liveness.go#L797)):*
> ```Go
> 
> 	if oldLiveness == (kvserverpb.Liveness{}) {
> 		// We don't yet know about our own liveness record (which does exist, we
> ```

Migration concerns, though? Makes me nervous. Please fall back to the manual initialization with a warning (instead of `ErrNoLivenessRecord`) to avoid trouble.
___
*[pkg/kv/kvserver/node_liveness.go, line 828 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGrEQeSD7tGGflj42lB:-MGrEQeSD7tGGflj42lC:b-87tib5) ([raw file](https://github.com/cockroachdb/cockroach/blob/64c3b57ebab29e59c19bf16f1c40bffc8591f8a4/pkg/kv/kvserver/node_liveness.go#L828)):*
> ```Go
> 	if incrementEpoch {
> 		newLiveness.Epoch++
> 		newLiveness.Draining = false // Clear draining field.
> ```

nit: `// clear draining field`

**Review by irfansharif — COMMENTED** [review]
Gotcha, that may have been FUD on my part. YOLO.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-:-MGsjO2498rAltIlyoTk:b-l0uesg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/node_liveness.go, line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGrE42x0TpH7S8PtOoG:-MGrGU_rDJBLEYOT8_Nm:b-x0j8z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/64c3b57ebab29e59c19bf16f1c40bffc8591f8a4/pkg/kv/kvserver/node_liveness.go#L797)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Migration concerns, though? Makes me nervous. Please fall back to the manual initialization with a warning (instead of `ErrNoLivenessRecord`) to avoid trouble.
</blockquote></details>

I don't think there are any migration concerns. By this phrasing I meant that this NodeLiveness guy hasn't yet discovered its liveness record (so the in-memory cache of liveness records needs updating).

In fact, if we did manually initialize here, it would actually do the same thing as we're doing explicitly. If we declared our oldLiveness to be empty, in trying to create a newLiveness our CPut attempt on the record below would fail. That would end up updating our in-memory cache with what was already found within KV, after which we'd have to retry anyway. I think writing it out in the way I did is a bit clearer (to me).

Separately, I think we actually want the error here to act as an assertion. Between us persisting a record during bootstrap, through the join rpc, and gossip code paths, there really should never be a missing liveness record once we get here. I'm inclined to let this bake on master for a bit as is to see what shakes out, and backporting to 20.2 after.

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/kv/kvserver/node_liveness.go, line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53842#-MGrE42x0TpH7S8PtOoG:-MGtashr3TINxxzP-kOo:b-83tw3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/64c3b57ebab29e59c19bf16f1c40bffc8591f8a4/pkg/kv/kvserver/node_liveness.go#L797)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I don't think there are any migration concerns. By this phrasing I meant that this NodeLiveness guy hasn't yet discovered its liveness record (so the in-memory cache of liveness records needs updating).

In fact, if we did manually initialize here, it would actually do the same thing as we're doing explicitly. If we declared our oldLiveness to be empty, in trying to create a newLiveness our CPut attempt on the record below would fail. That would end up updating our in-memory cache with what was already found within KV, after which we'd have to retry anyway. I think writing it out in the way I did is a bit clearer (to me).

Separately, I think we actually want the error here to act as an assertion. Between us persisting a record during bootstrap, through the join rpc, and gossip code paths, there really should never be a missing liveness record once we get here. I'm inclined to let this bake on master for a bit as is to see what shakes out, and backporting to 20.2 after.
</blockquote></details>

If a 20.1 node joins the cluster (at 20.1) and gets killed before persisting its liveness record, and then the cluster starts running 20.2 binaries, and the node comes back with the code of this PR, won't it eternally be unable to heartbeat itself? It's not going to get its liveness record created by the join rpc (it's not joining) nor bootstrap. I agree that this is sort of rare but if it's possible - we ought to handle it. What am I missing?

**Comment by irfansharif** [comment]
Good catch, hopefully the last time we (I) have to think about missing liveness records. Added it in #54216.

**Pull request merged**
